rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,518 @@
+
+Here is an example of how the format used in *nixdoc* works:
+
+```nix
+#attrsets. nix (simplified)
+
+{ lib }:
+# Operations on attribute sets.
+
+let
+# ...
+in
+{
+  /* <Description Field>
+     Example:
+     <Comprehensive Code example>
+     Type:
+     <Type signature>
+  */
+  AttrFunc =
+    # <Desribe arg1>
+    arg1:
+    # <Describe arg2>
+    arg2:
+    
+    # ... implementation
+}
+```
+
+## Drawbacks 
+
+##### Unspecified format
+
+In general, the format for writing documentation strings is **not specified**. The only place where it is applied is: *nixpkgs/lib/**
+
+*nixdoc* only applies to places in nixpkgs/lib. But extending the scope of *nixdoc* is not the primary goal. Instead, we should find formal rules for writing *doc-strings*. Tools like *nixdoc* can then implement against this RFC instead of the format relying on nixdoc implementation details.
+
+##### Only specific places
+
+The placement of those comments requires precisely looking at the attribute set containing the function declarations, which is not usable for general-purpose documentation strings. 
+
+e.g., 
+
+- file that directly exports the lib-function without wrapping it in an attribute set.
+- file that exports a constant
+- files outside of lib are not / cannot be rendered
+
+##### Differentiate from regular comments
+
+The format doesn't allow any distinction between doc-strings or regular comments.
+
+Having a distinction would allow us to
+
+1. Find all comments that are part of the documentation
+2. Render them in the documentation format
+3. Connect the documentation to the exact places in the nix code. This is already done, but only for nixpkgs/lib.
+
+##### References 
+
+###### nixpkgs - Dosctrings examples
+
+- [lib/attrsets](https://github.com/NixOS/nixpkgs/blob/master/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix)
+- more...
+
+###### frameworks
+
+- [dream2nix/utils](https://github.com/nix-community/dream2nix/blob/main/src/utils/config.nix)
+- [dream2nix/templates/builder](https://github.com/nix-community/dream2nix/blob/main/src/templates/builders/default.nix)
+- more... 
+
+###### Other tools
+
+Other tools that work directly with the nix AST and comments:
+
+- [noogle](https://noogle.dev) - Nix API search engine. It allows you to search functions and other expressions.
+- [nix-doc](https://github.com/lf-/nix-doc) - A Nix developer tool leveraging the rnix Nix parser for intelligent documentation search and tags generation
+- [manix](https://github.com/mlvzk/manix) - A fast CLI documentation searcher for nix. 
+
+## Proposed Improvement - Doc-blocks
```

Comment:
The sections in this RFC are very confusing. You had a "Drawbacks" section earlier, but it's about drawbacks to the pre-RFC state. I'd rather call that "Current problems" or so.

And here you're talking about a proposed improvement, but we're still in the overall "Motivation" section. The actual proposal should be in the "Detailed design" section instead.

You also have an "Alternatives" section that comes before the "Detailed design" section, that's backwards.

All of this makes the RFC rather hard to read and understand.

Reply:
The reason i went with the alternatives section first is because i wanted to open the discussion about the different possibles outcomes first. Before proposing my personal favorite.

Do you think this makes sense?

Reply:
I switched the sections back to the order in the template. You are right; it is confusing that way.